htlcswitch/switch.go

```diff
@@ -353,35 +340,80 @@ func (s *Switch) ProcessContractResolution(msg contractcourt.ResolutionMsg) erro
 	return nil
 }
 
+// GetPaymentResult returns the the result of the payment attempt with the
+// given paymentID. The method returns a channel where the payment result will
+// be sent when available, or an error is encountered. If the paymentID is
+// unknown, ErrPaymentIDNotFound will be returned.
+func (s *Switch) GetPaymentResult(paymentID uint64,
+	deobfuscator ErrorDecrypter) (<-chan *PaymentResult, error) {
+
+	s.pendingMutex.Lock()
+	payment, ok := s.pendingPayments[paymentID]
+	s.pendingMutex.Unlock()
+
+	if !ok {
+		return nil, ErrPaymentIDNotFound
+	}
+
+	resultChan := make(chan *PaymentResult, 1)
+
+	// Since the payment was known, we can start a goroutine that can
+	// extract the result when it is available, and pass it on to the
+	// caller.
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		var n *networkResult
+		select {
+		case n = <-payment.resultChan:
+		case <-s.quit:
+			return
+		}
+
+		// Extract the result and pass it to the result channel.
+		result, err := s.extractResult(
+			deobfuscator, n, payment.paymentHash,
+		)
+		if err != nil {
+			log.Errorf("Unable to extract result: %v", err)
+			return
```

Comment:
This will leave the `router` blocked?

Reply:
Yeah we need to return an error result in this case. Also above when we exit due to a quit channel.

Reply:
Hm, yeah this is interesting. When calling `GetPaymentResult` we need to distinguish between 3 types of errors:

1. An immediate error whether the subscription was successful (currently only error is in case of unknown payment ID).
2. Critical error that happened when the result was received (this case). This is not a `ForwardingError` (though we could wrap it in one?).
3. A forwarding error.

I see two ways of designing this:

* Make `PaymentResult` also hold critical errors:
```go
type PaymentResult struct {
	// Preimage is set by the switch in case a sent HTLC was settled.
	Preimage [32]byte

 	// Error is non-nil in case a HTLC send failed, and the HTLC is now
	// irrevocably cancelled.
	Error error
}
```
The reason this could only be `ForwardingErrors` is that in an earlier iteration this whole struct was serialized.
* Wrap the critical error in a `ForwardingError`.

Reply:
I realized that we cannot return any result in case of quit, so maybe we can just close the channel in that case.

Reply:
Made a thing: https://github.com/lightningnetwork/lnd/pull/3063/commits/1fa25c266a357dae5f6d758a968df44b5d2eacce